### PR TITLE
adding the option to use a keyfile to unlock the device

### PR DESCRIPTION
**Pull request by Keksgesicht**
This should solve issue #20961 and #23295.

Does someone also need `--keyfile-offset` and `--keyfile-size`?
You could create a new file, fill it with random data, add it via `cryptsetup` to the volume and use this one instead.

For the case of stdin just use `--unlock-key-file=/dev/stdin`.

If I somehow created a memory leak or have to sanitize an input just let me know.

### CODE REVIEW

File: src/cryptenroll/cryptenroll.c [hunk 1]
```diff
@@ -221,6 +226,10 @@ static int parse_argv(int argc, char *argv[]) {
                         arg_enroll_type = ENROLL_RECOVERY;
                         break;
 
+                case ARG_UNLOCK_FILE:
+                        arg_unlock_file = strdup(optarg);
```
Comment (bluca) [inline src/cryptenroll/cryptenroll.c]: OOM check missing

File: src/cryptenroll/cryptenroll.c [hunk 2]
```diff
@@ -486,6 +495,30 @@ static int prepare_luks(
                                 strlen(envpw));
                 if (r < 0)
                         return log_error_errno(r, "Password from environment variable $PASSWORD did not work.");
+        } else if(arg_unlock_file) {
+                char *password;
```
Comment (bluca) [inline src/cryptenroll/cryptenroll.c]: `_cleanup_(erase_and_freep) char *password = NULL`

File: src/cryptenroll/cryptenroll.c [hunk 3]
```diff
@@ -486,6 +495,30 @@ static int prepare_luks(
...
+
+                r = crypt_volume_key_get(
+                                cd,
+                                CRYPT_ANY_SLOT,
+                                vk,
+                                &vks,
+                                password,
+                                passwordLen);
+
+                crypt_safe_free(password);
```
Comment (bluca) [inline src/cryptenroll/cryptenroll.c]: not needed

File: src/cryptenroll/cryptenroll.c [hunk 4]
```diff
@@ -486,6 +495,30 @@ static int prepare_luks(
...
+                if (r < 0)
+                        return log_error_errno(r, "Unlocking via keyfile did not work.");
```
Comment (bluca) [inline src/cryptenroll/cryptenroll.c]: `Unlocking via keyfile failed: %m`

File: src/cryptenroll/cryptenroll.c [hunk 5]
```diff
@@ -486,6 +495,30 @@ static int prepare_luks(
...
+                size_t passwordLen;
+
+                r = crypt_keyfile_device_read(
+                                cd,
+                                arg_unlock_file,
+                                &password,
+                                &passwordLen,
+                                0, 0, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Unlocking via keyfile did not work.");
```
Comment (bluca) [inline src/cryptenroll/cryptenroll.c]: `Reading keyfile failed: %m`

File: src/cryptenroll/cryptenroll.c [hunk 6]
```diff
@@ -486,6 +495,30 @@ static int prepare_luks(
                                 strlen(envpw));
                 if (r < 0)
                         return log_error_errno(r, "Password from environment variable $PASSWORD did not work.");
+        } else if(arg_unlock_file) {
+                char *password;
+                size_t passwordLen;
```
Comment (bluca) [inline src/cryptenroll/cryptenroll.c]: nit: we don't do camel case

**Review by bluca — CHANGES_REQUESTED** [review]
Please update the documentation, and also add a simple test case to test/units/testsuite-70.sh

File: src/cryptenroll/cryptenroll.c [hunk 7]
```diff
@@ -28,6 +28,7 @@
 #include "tpm2-util.h"
 
 static EnrollType arg_enroll_type = _ENROLL_TYPE_INVALID;
+static char *arg_unlock_file = NULL;
```
Comment (bluca) [inline src/cryptenroll/cryptenroll.c]: needs a `STATIC_DESTRUCTOR_REGISTER` a bit below

File: src/cryptenroll/cryptenroll.c (hunk 2 above)
Comment (bluca) [reply src/cryptenroll/cryptenroll.c]: still not addressed

Comment (Keksgesicht) [reply src/cryptenroll/cryptenroll.c]: changing this gives me:
```
free(): invalid pointer
Aborted
```
then `--unlock-file` parameter is provided.

File: src/cryptenroll/cryptenroll.c [hunk 8]
```diff
@@ -486,6 +499,28 @@ static int prepare_luks(
...
+                size_t password_len;
+
+                r = crypt_keyfile_device_read(
+                                cd,
+                                arg_unlock_file,
+                                &password,
+                                &password_len,
+                                0, 0, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Unlocking via keyfile failed: %m");
```
Comment (bluca) [inline src/cryptenroll/cryptenroll.c]: this is not unlocking, it's reading, so as mentioned: `"Reading keyfile %s failed: %m", arg_unlock_file`

File: src/cryptenroll/cryptenroll.c (hunk 2 above)
Comment (bluca) [reply src/cryptenroll/cryptenroll.c]: bah, it looks like it's not an actual string, it's a hidden special structure. You'll need to add a build time check for crypt_safe_free, as it's available in newer versions than the minimum requirement, and wrap the whole block in an ifdef and use crypt_safe_free

File: src/cryptenroll/cryptenroll.c (hunk 7 above)
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: hmm, so crypttab and cryptsetup call the concept "keyfile". We should probably stick to the same naming, to minimize surprises.

File: src/cryptenroll/cryptenroll.c [hunk 9]
```diff
@@ -221,6 +227,13 @@ static int parse_argv(int argc, char *argv[]) {
                         arg_enroll_type = ENROLL_RECOVERY;
                         break;
 
+                case ARG_UNLOCK_FILE:
+                        arg_unlock_file = strdup(optarg);
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: please use parse_path_argument()

File: src/cryptenroll/cryptenroll.c [hunk 10]
```diff
@@ -486,6 +499,28 @@ static int prepare_luks(
                                 strlen(envpw));
                 if (r < 0)
                         return log_error_errno(r, "Password from environment variable $PASSWORD did not work.");
+        } else if(arg_unlock_file) {
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: space between `if` and `(`...

More importantly though. I am pretty sure if the user specified the file on the cmdline, it should take precedence over the PASSWORD env var. I.e. do this stuff first.

File: src/cryptenroll/cryptenroll.c [hunk 11]
```diff
@@ -486,6 +499,28 @@ static int prepare_luks(
...
+                char *password;
+                size_t password_len;
+
+                r = crypt_keyfile_device_read(
+                                cd,
+                                arg_unlock_file,
+                                &password,
+                                &password_len,
+                                0, 0, 0);
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: please use read_full_file() or read_full_file_full() and use `READ_FULL_FILE_SECURE|READ_FULL_FILE_WARN_WORLD_READABLE|READ_FULL_FILE_CONNECT_SOCKET`

File: src/cryptenroll/cryptenroll.c (hunk 2 above)
Comment (poettering) [reply src/cryptenroll/cryptenroll.c]: dont, just use read_full_file() see below, and then use `_cleanup_(erase_and_freep)`

Comment (poettering) [reply src/cryptenroll/cryptenroll.c]: also, given this is now binary (i.e. not a NUL-terminated string anymore), please use `void*`, and not `char*` (which suggests this was a string)

**Comment by poettering** [comment]
(btw, i figure we should also allow enrolling key files with cryptenroll eventually, and then set a hint on the key slot indicating if it contains non-printable chars, as then it makes little sense to ask for the key interactively at boot)

File: src/cryptenroll/cryptenroll.c (hunk 7 above)
Comment (Keksgesicht) [reply src/cryptenroll/cryptenroll.c]: Just to be sure. In https://github.com/systemd/systemd/issues/23295#issuecomment-1122119881 you said `--unlock-file=` make sense to you.
So are we using `--unlock-file=` and `--key-file=` or `--key-file=` and an unnamed parameter with the keyfile path like `cryptsetup`?

Comment (poettering) [reply src/cryptenroll/cryptenroll.c]: uh, oh. thanks for pointing me to my own earlier contradicting words ;-)

Hmm, so I think my current self would agree with my younger self there, i.e. that this probably should be called something with "unlock" to indicate that this is about unlocking for enrollment, not for enrolling. But my current self also has a point, it should use existing nomenclature. So maybe the combination of both make most sense, i.e. `--unlock-key-file=…`. That way we have the best of both worlds, and call it "key file" but also emphasize it is about "unlocking".

Does that make sense?

**Comment by poettering** [comment]
> (btw, i figure we should also allow enrolling key files with cryptenroll eventually, and then set a hint on the key slot indicating if it contains non-printable chars, as then it makes little sense to ask for the key interactively at boot)

(And yes, apparently my younger self disagree with that opinion ;-))

File: src/cryptenroll/cryptenroll.c (hunk 7 above)
Comment (medhefgo) [reply src/cryptenroll/cryptenroll.c]: So this isn't your usual keyfile (e.g. binary blob) we are used to from cryptsetup operations, but rather a file containing the password (which I think is restricted in content to some degree?), right? In that case it should really be called password_file or something like that to drive that point home.

Comment (poettering) [reply src/cryptenroll/cryptenroll.c]: Hmm, it isn't you usual key file? huh?

Comment (medhefgo) [reply src/cryptenroll/cryptenroll.c]: My cursory glance at this was that this option is just a means to pass the password to sd-cryptentroll (non-interactively). Whereas a keyfile would be a binary blob and something else entirely. I might be wrong here, though.

Comment (poettering) [reply src/cryptenroll/cryptenroll.c]: cryptsetup/LUKS do not distinguish passwords from binary blobs used to unlock stuff. It's always just some data with a size...

Comment (medhefgo) [reply src/cryptenroll/cryptenroll.c]: I would beg to differ (yes, the low-level stuff doesn't care, but the UI should/does): https://man.archlinux.org/man/cryptsetup.8#Passphrase_processing_for_LUKS

Comment (poettering) [reply src/cryptenroll/cryptenroll.c]: I don't grok what you are trying to say?

Comment (Keksgesicht) [reply src/cryptenroll/cryptenroll.c]: > So maybe the combination of both make most sense, i.e. `--unlock-key-file=…`. That way we have the best of both worlds, and call it "key file" but also emphasize it is about "unlocking".

This sounds like an good idea.

**Comment by poettering** [comment]
lgtm

**Comment by MrMarlin** [comment]
Hi. Thanks for adding this functionality. When will this be available in a release?

**Comment by josephlr** [comment]
> Hi. Thanks for adding this functionality. When will this be available in a release?

Looking at https://github.com/systemd/systemd/commit/38db7a4ed388b032faae221067b77d553a54e6a6 it seems like this is in v252 (and later).

**Pull request merged**